Rewrite WebCodecs blog post (#6015)
GoogleChrome/web.dev: src/site/content/en/blog/webcodecs/index.md

## Current status {: #status }

The [WebCodecs API](https://wicg.github.io/web-codecs/) is available in Chrome 94.

## Video frames

Frames are the centerpiece in video processing. Thus in WebCodecs most classes
either consume or produce frames. Video encoders convert frames into encoded
chunks. Video decoders do the opposite.

Also `VideoFrame` plays nicely with the the rest of Web APIs by being a [`CanvasImageSource`](https://html.spec.whatwg.org/multipage/canvas.html#canvasimagesource) and having a [constructor](https://www.w3.org/TR/webcodecs/#dom-videoframe-videoframe) accepting `CanvasImageSource`.
So it can be used in functions like [`drawImage()`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage) and[`texImage2D()`](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texImage2D), also it can be constructed from canvases, bitmaps, video elements and other video frames.

> WebCodecs API works best in tandem with the classes from [Insertable Streams API](https://w3c.github.io/mediacapture-transform/)

Reviewer: Nit: I would prefer "works well"

To help with that, [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)
provides a convenient way to automatically transfer all frames coming from a media
track to the worker. For example, `MediaStreamTrackProcessor` can be used to obtain a
`ReadableStream` for a media stream track coming from the web camera, after that
the stream is transferred to a web worker where frames are read one by one and queued
into a `VideoEncoder`.

With [`HTMLCanvasElement.transferControlToOffscreen`](https://developers.google.com/web/updates/2018/08/offscreen-canvas#unblock_main_thread) even rendering can be done off the main thread. But if all the hight level tools turned
out to be inconvenient, `VideoFrame` itself is [transferable](https://developer.mozilla.org/en-US/docs/Web/API/Transferable) and may be
moved between workers.

## WebCodecs in action

### Encoding

It all starts with a `VideoFrame`.
There are three ways to construct video frames
1. From an image source like canvas, image bitmap or a video element.
```js
const cnv = document.createElement('canvas');
// draw something on the canvas
…
let frame_from_canvas = new VideoFrame(cnv, { timestamp: 0 });
```

> 2. Use `MediaStreamTrackProcessor` to pull frames from a [`MediaStreamTrack`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack)

Reviewer: This isn't rendering correctly because it doesn't know this is the same numbered list.

```js
  const stream = await navigator.mediaDevices.getUserMedia({ … });
  const track = stream.getTracks()[0];

  const media_processor = new MediaStreamTrackProcessor(track);

  const reader = media_processor.readable.getReader();
  while (true) {
      const result = await reader.read();
      if (result.done)
        break;
      let frame_from_camera = result.value;
  }
```

> 3. Create a frame from its binary pixel representation in a [`BufferSource`](https://developer.mozilla.org/en-US/docs/Web/API/BufferSource)

Reviewer: Probably needs a body/example, right now it looks like the figures below are the explanation.

which connect WebCodecs to [media stream tracks](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack).
- `MediaStreamTrackProcessor` breaks up into media tracks into individual frames
- `MediaStreamTrackGenerator` helps to create a media track from a stream of frames.

## WebCodecs and web workers

By design WebCodecs API does all the heavy lifting asynchronously and off the main thread,
but since frame and chunk callbacks can often be called multiple times a second,
they might clutter the main thread and thus make the website less responsive.
Therefore it is preferable to move handling of individual frames and encoded chunks into a
web worker.

To help with that, [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)
provides a convenient way to automatically transfer all frames coming from a media
track to the worker. For example, `MediaStreamTrackProcessor` can be used to obtain a
`ReadableStream` for a media stream track coming from the web camera, after that
the stream is transferred to a web worker where frames are read one by one and queued
into a `VideoEncoder`.

> With [`HTMLCanvasElement.transferControlToOffscreen`](https://developers.google.com/web/updates/2018/08/offscreen-canvas#unblock_main_thread) even rendering can be done off the main thread. But if all the hight level tools turned

Author: The article that I linked has details about usage of the offscreen canvas, in particular how to use it from a worker 

```js
const offscreen = document.querySelector('canvas').transferControlToOffscreen();
const worker = new Worker('myworkerurl.js');
worker.postMessage({ canvas: offscreen }, [offscreen]);
```

> With [`HTMLCanvasElement.transferControlToOffscreen`](https://developers.google.com/web/updates/2018/08/offscreen-canvas#unblock_main_thread) even rendering can be done off the main thread. But if all the hight level tools turned

Reviewer 2: I've opened a PR to add this. https://github.com/mdn/content/pull/8265

> With [`HTMLCanvasElement.transferControlToOffscreen`](https://developers.google.com/web/updates/2018/08/offscreen-canvas#unblock_main_thread) even rendering can be done off the main thread. But if all the hight level tools turned

Reviewer 2: Even if merged, the MDN PR won't be deployed until after this needs to go out.

out to be inconvenient, `VideoFrame` itself is [transferable](https://developer.mozilla.org/en-US/docs/Web/API/Transferable) and may be
moved between workers.

## WebCodecs in action

### Encoding

```html

>   <Img src="Encoding.png" alt="The path from a Canvas or an ImageBitmap to the network or to storage" width="960" height="720" />

Reviewer 2: ```suggestion
  <img src="Encoding.png" alt="The path from a Canvas or an ImageBitmap to the network or to storage" width="960" height="720" />
```

Reviewer 2: Is this an example you're working with in the following section? I don't see how it relates since it's not a video.

Author: I just marked an image that needs to be used here as an illustration.

This kinda stuff didn't work for me, because I don't know how to upload imaged to the CDN
<figure class="w-figure">
  {% Img src="image/admin/7LNdjNlwCUB1csF9DDl8.png", alt="The path from ImageBitmap to the network or to storage.", width="800", height="291" %}
  <figcaption class="w-figcaption">The path from <code>ImageBitmap</code> to the network or to storage.</figcaption>
</figure>

```

It all starts with a `VideoFrame`.

> There are three ways to construct video frames

Reviewer 2: ```suggestion
There are three ways to construct video frames.
```

> + From an image source like canvas, image bitmap or a video element.

Reviewer 2: ```suggestion
+ From an image source like a canvas, an image bitmap, or a video element.
```

which connect WebCodecs to [media stream tracks](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack).
- `MediaStreamTrackProcessor` breaks media tracks into individual frames.
- `MediaStreamTrackGenerator` creates a media track from a stream of frames.

## WebCodecs and web workers

By design WebCodecs API does all the heavy lifting asynchronously and off the main thread.
But since frame and chunk callbacks can often be called multiple times a second,
they might clutter the main thread and thus make the website less responsive.
Therefore it is preferable to move handling of individual frames and encoded chunks into a
web worker.

To help with that, [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream)
provides a convenient way to automatically transfer all frames coming from a media
track to the worker. For example, `MediaStreamTrackProcessor` can be used to obtain a
`ReadableStream` for a media stream track coming from the web camera. After that
the stream is transferred to a web worker where frames are read one by one and queued
into a `VideoEncoder`.

> With [`HTMLCanvasElement.transferControlToOffscreen`](https://developers.google.com/web/updates/2018/08/offscreen-canvas#unblock_main_thread) even rendering can be done off the main thread. But if all the hight level tools turned

Reviewer 3: Spelling error here: `hight` should be `high`

out to be inconvenient, `VideoFrame` itself is [transferable](https://developer.mozilla.org/en-US/docs/Web/API/Transferable) and may be
moved between workers.

## WebCodecs in action

### Encoding

<figure class="w-figure">

>   {% Img src="Encoding.png", alt="The path from a Canvas or an ImageBitmap to the network or to storage", width="960", height="720" %}

Reviewer 2: ```suggestion
  {% Img src="image/sQ51XsLqKMgSQMCZjIN0B7hlBO02/lEovMYp8oh1JSClCLCiD.png", alt="The path from a Canvas or an ImageBitmap to the network or to storage", width="800", height="393" %}
```

  codec: 'vp8',
  width: 640,
  height: 480,
  bitrate: 2_000_000,     // 2 Mbps
  framerate: 30,
};

let encoder = new VideoEncoder(init);
encoder.configure(config);
```

After the encoder has been set up, it's ready to start accepting frames via `encode()` method.
Both `configure()` and `encode()` return immediately without waiting for the
actual work to complete. It allows several frames to queue for encoding at the
same time, `encodeQueueSize()` shows how many requests are waiting in the queue waiting
for previous encodes to finish.

> Errors are reported either by immediately throwing exceptions or by calling the `error()`

Reviewer: > Some errors are easy to detect immediately

Possibly misleading. Ease of detection is not our criteria, layering is. Errors are thrown immediately for violations of the API contract, `error()` is called for problems encountered in the codec implementation.

  codec: 'vp8',
  width: 640,
  height: 480,
  bitrate: 2_000_000, // 2 Mbps
  framerate: 30,
};

let encoder = new VideoEncoder(init);
encoder.configure(config);
```

> After the encoder has been set up, it's ready to start accepting frames via `encode()` method.

Reviewer 2: ```suggestion
After the encoder has been set up, it's ready to accept frames via `encode()` method.
```

Reviewer 2: There's a writing best practice to omit needless words, that is, words that don't contribute meaning. A common source is what's called an inflated verb. If there are two verb forms ("start accepting") it can almost always be compressed to the second (accept).

Both `configure()` and `encode()` return immediately without waiting for the
actual work to complete. It allows several frames to queue for encoding at the

> same time, `encodeQueueSize` shows how many requests are waiting in the queue waiting

Reviewer 2: ```suggestion
same time, while `encodeQueueSize` shows how many requests are waiting in the queue
```

Reviewer 2: The alternative to "while" is to make `encodeQueueSize` the start of another sentence.

for previous encodes to finish.

> Errors are reported either by immediately throwing an exception, if case the arguments

Reviewer 2: ```suggestion
Errors are reported either by immediately throwing an exception, in case the arguments
```

or the order of method calls violates the API contract, or by calling the `error()`
callback for problems encountered in the codec implementation.
If encoding completes successfully the `output()`
callback is called with a new encoded chunk as an argument.

> Another important detail here is that frames need to be told when they are nor

Reviewer 2: ```suggestion
Another important detail here is that frames need to be told when they are no
```

container for storage.

```js
function handleChunk(chunk, metadata) {

  if (metadata.decoderConfig) {
    // Decoder needs to be configured (or reconfigured) with new parameters
    // when metadata has a new decoderConfig.
    // Usually it happens in the beginning or when encoder has a new codec
    // extra data (aka decoderConfig.description).

    fetch('/upload_extra_data',
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/octet-stream' },
      body: metadata.decoderConfig.description
    });
  }

  // actual bytes of encoded data

>   let chunkData = new Uint8Array(chunk.byteLength);

Reviewer: `fetch()` below is still using `data`

container for storage.

```js
function handleChunk(chunk, metadata) {

  if (metadata.decoderConfig) {
    // Decoder needs to be configured (or reconfigured) with new parameters
    // when metadata has a new decoderConfig.
    // Usually it happens in the beginning or when encoder has a new codec

>     // extra data (aka decoderConfig.description).

Reviewer 2: >a new codec extra data

What?

Setting up a `VideoDecoder` is similar to what's been done for the
`VideoEncoder`: two functions are passed when the decoder is created, and codec
parameters are given to `configure()`.

> The set of codec parameters can vary from codec to codec, for example H264 codec

Reviewer: H.264

might need a [binary blob](https://wicg.github.io/web-codecs/#dom-audiodecoderconfig-description)

> of AVCC extradata, unless it's encoded in so called AnnexB format ( `encoderConfig.avc = { format: "annexb" }` ).

Reviewer: avcC. I'm not sure if we should call it extradata which is an FFmpeg-ism, bit it may be in common-enough usage to be helpful.

### Decoding

```html
  <Img src="Decoding.png" alt="The path from the network or storage to a Canvas or an ImageBitmap." width="960" height="720" />

> ```

Reviewer 2: Are these supposed to be images shown in the article?

Author: yes, <figure> tag didn't work for me :(

Reviewer 2: This won't work either. I have to upload them somewhere and give you the links. Can you please put the figure tag back in.

Your images have a lot of extra whitespace. Do you mind if I crop them?

Author: I brought back the figure tags. Feel free to crop or ever redraw images.

Setting up a `VideoDecoder` is similar to what's been done for the
`VideoEncoder`: two functions are passed when the decoder is created, and codec
parameters are given to `configure()`.

> The set of codec parameters can vary from codec to codec, for example H.264 codec

Reviewer 2: ```suggestion
The set of codec parameters varies from codec to codec. For example H.264 codec
```

decoder.configure(config);
```

Once the decoder is initialized, you can start feeding it with `EncodedVideoChunk` objects.

> In order to create a chunk

Reviewer: Something seems to have gone missing here.

decoder.configure(config);
```

Once the decoder is initialized, you can start feeding it with `EncodedVideoChunk` objects.

> In order to create a chunk, you'll need

Reviewer 2: ```suggestion
To create a chunk, you'll need:
```

> - a [`BufferSource`](https://developer.mozilla.org/en-US/docs/Web/API/BufferSource)of encoded video data

Reviewer 2: ```suggestion
- A [`BufferSource`](https://developer.mozilla.org/en-US/docs/Web/API/BufferSource) of encoded video data
```

- the chunk's start timestamp in microseconds (media time of the first encoded frame in the chunk)

> - the chunk's type

Reviewer 2: ```suggestion
- the chunk's type, one of:
```

  - `key` if the chunk can be decoded independently from previous chunks
  - `delta` if the chunk can only be decoded after one or more previous chunks have been decoded

Any chunks emitted by the encoder are ready for the decoder as is, although it's

> hard to imagine a real-world use case for decoding newly encoded chunks (except for the demo below).

Reviewer 2: Debugging a video conferencing or video editing app?

  let frame = ready_frames.shift();
  underflow = false;

  // Based on the frame's timestamp calculate how much of real time waiting
  // is needed before showing the next frame.
  let time_till_next_frame = calculateTimeTillNextFrame(frame.timestamp);
  await delay(time_till_next_frame);
  ctx.drawImage(frame, 0, 0);
  frame.close();

  // Immediately schedule rendering of the next frame
  setTimeout(render_frame, 0);
}
```

## Demo

The demo below shows how animation frames from a canvas are
- captured at 25fps into a `ReadableStream` by `MediaStreamTrackProcessor`
- transferred to a web worker
- encoded into H.264 video format
- decoded again into a sequence of video frames
- and rendered on the second canvas using `transferControlToOffscreen()`

{% Glitch 'new-webcodecs-blogpost-demo' %}

## Using the WebCodecs API {: #use }

## Feature detection

To check for WebCodecs support:

```js
if ('VideoFrame' in window) {
  // WebCodecs API is supported.
}
```

## Feedback {: #feedback }

The Chrome team wants to hear about your experiences with the WebCodecs API.

> ### Tell us about the API design

Reviewer: I don't think we're soliciting feedback on the overall design any more, we're on to v2 feature requests now.

Author: It's predefined boilerplate text from web.dev

  let frame = ready_frames.shift();
  underflow = false;

  // Based on the frame's timestamp calculate how much of real time waiting
  // is needed before showing the next frame.
  let time_till_next_frame = calculateTimeTillNextFrame(frame.timestamp);
  await delay(time_till_next_frame);
  ctx.drawImage(frame, 0, 0);
  frame.close();

  // Immediately schedule rendering of the next frame
  setTimeout(render_frame, 0);
}
```

## Demo

> The demo below shows how animation frames from a canvas are

Reviewer 2: ```suggestion
The demo below shows how animation frames from a canvas are:
```
